frame/dex-router/src/lib.rs

```diff
@@ -242,43 +402,117 @@ pub mod pallet {
 				)?;
 				dx_t = dy_t;
 			}
+			ensure!(dy_t >= min_receive, Error::<T>::CanNotRespectMinAmountRequested);
 			Ok(dy_t)
 		}
 
+		#[transactional]
 		fn sell(
 			who: &T::AccountId,
 			asset_pair: CurrencyPair<T::AssetId>,
 			amount: T::Balance,
+			min_receive: T::Balance,
 		) -> Result<T::Balance, DispatchError> {
-			Self::exchange(who, asset_pair, amount)
+			<Self as DexRouter<
+				T::AccountId,
+				T::AssetId,
+				T::PoolId,
+				T::Balance,
+				T::MaxHopsInRoute,
+			>>::exchange(who, asset_pair, amount, min_receive)
 		}
 
+		#[transactional]
 		fn buy(
 			who: &T::AccountId,
 			asset_pair: CurrencyPair<T::AssetId>,
 			amount: T::Balance,
+			min_receive: T::Balance,
 		) -> Result<T::Balance, DispatchError> {
-			let route = Self::get_route(asset_pair).ok_or(Error::<T>::NoRouteFound)?;
+			let (route, reverse) = Self::get_route(asset_pair).ok_or(Error::<T>::NoRouteFound)?;
 			let mut dy_t = amount;
 			let mut dx_t = T::Balance::zero();
-			for pool_id in route.iter().rev() {
-				let currency_pair = T::Pablo::currency_pair(*pool_id)?;
+			let mut forward_iter;
+			let mut backward_iter;
+			let route_iter: &mut dyn Iterator<Item = &T::PoolId> = if !reverse {
```

Comment:
i believe you may do it withoyt dyn

Reply:
I tried but it does not work. Is it prohibited to use `dyn` in our codebase? if yes could you please explain the details.